src/cachetools/__init__.py

```diff
@@ -497,49 +502,73 @@ def __getlink(self, key):
         return value
 
 
+def _get_decorator_value_from_cache(cache, key, lock=None):
+    """Returns the cached value for the given `key` from the given `cache`
+    instance.
+
+    In case the key is not found in the cache, a sentinel value is returned.
+    This is to avoid to reuse None as a "not found" marker since None is a
+    valid return value for decorated functions (in some cases it might take
+    an expensive computation to return None)."""
+    lock = lock or contextlib.nullcontext()
+    try:
+        with lock:
+            return cache[key]
+    except KeyError:
+        return _DECORATOR_SENTINEL
+
+
+def _set_decorator_value_in_cache(cache, key, value, lock=None):
+    lock = lock or contextlib.nullcontext()
+    try:
+        with lock:
+            cache.setdefault(key, value)
+    except ValueError:
+        pass
+
+
 def cached(cache, key=hashkey, lock=None):
     """Decorator to wrap a function with a memoizing callable that saves
     results in a cache.
 
     """
 
     def decorator(func):
-        if cache is None:
+        if inspect.iscoroutinefunction(func):
```

Comment:
Assuming we want to support callable objects, this check wouldn't be enough. 

First, we might want to consider using 
`asyncio.iscoroutinefunction` instead of `inspect.iscoroutinefunction` since the function from asyncio handles functions decorated with ` @coroutine` as well, although we might just say there's no support for legacy coroutines for a small performance boost in this check (https://docs.python.org/3/library/asyncio-task.html#asyncio.iscoroutinefunction, https://github.com/python/cpython/blob/3.10/Lib/asyncio/coroutines.py#L164).

That's still not enough to handle callable objects though.

So for example
```python
class AsyncCallable:
    async def __call__():
        pass

assert not inspect.iscoroutinefunction(AsyncCallable())    # This is not a coroutine function
```

To make this work we would need to modify the check to something like:
```python
import asyncio
import inspect
from typing import Callable


def is_async_callable(c: Callable):
    return (
            asyncio.iscoroutinefunction(c) or
            (not inspect.isclass(c) and asyncio.iscoroutinefunction(getattr(c, "__call__", None)))
    )


def function_():
    pass


async def async_function():
    pass


lambda_ = lambda: None
# It's a fair assumption that fake "async lambdas" won't be supported

class NotCallable:
    pass


class SyncCallable:
    def __call__(self):
        pass


class AsyncCallable:
    async def __call__(self):
        pass


assert not is_async_callable(function_)
assert is_async_callable(async_function)
assert not is_async_callable(lambda_)
assert not is_async_callable(NotCallable())
assert not is_async_callable(SyncCallable())
assert is_async_callable(AsyncCallable())
assert not is_async_callable(NotCallable)
assert not is_async_callable(SyncCallable)
assert not is_async_callable(AsyncCallable)    # That's why we inspect if it's isclass


assert not asyncio.iscoroutinefunction(function_)
assert asyncio.iscoroutinefunction(async_function)
assert not asyncio.iscoroutinefunction(lambda_)
assert not asyncio.iscoroutinefunction(NotCallable())
assert not asyncio.iscoroutinefunction(SyncCallable())
assert not asyncio.iscoroutinefunction(AsyncCallable())
assert not asyncio.iscoroutinefunction(NotCallable)
assert not asyncio.iscoroutinefunction(SyncCallable)
assert not asyncio.iscoroutinefunction(AsyncCallable)
```

Reply:
> Assuming we want to support callable objects, this check wouldn't be enough.

That's not an assumption I made, since my personal code (the one that triggered my itch to write this PR) does not use callable objects. As you point out, changing the predicate would add support for callable objects quite trivially however.